File: my-folder/0498-diagonal-traverse/solution.java
class Solution {
    public int[] findDiagonalOrder(int[][] matrix) {
        int rows = matrix.length;
        if(rows == 0){
            return new int[0];
        }
        int cols = matrix[0].length;
        int total = rows * cols;
        int[] res = new int[total];
        
        int i=0;
        int j=0;
        for(int k=0;k<total;++k){
            res[k] = matrix[i][j];
            // even
            if((i+j)%2==0){ // up
                if(j==cols-1){
                    i++;
                }else if(i==0){
                    j++;
                }else{
                    i--;
                    j++;
                }
            }else{ // down
                if(i==rows-1){
                    j++;
                }else if(j==0){
                    i++;
                }else{
                    i++;
                    j--;
                }
            }
        }
        
        return res;
    }
}

File: my-folder/0498-diagonal-traverse/solution.py
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        ans = []
        m,n = len(mat), len(mat[0])
        for k in range(m + n - 1):
            t = []
            i = 0 if k < n else k - n + 1
            j = k if k < n else n - 1
            while i < m and j >= 0:
                t.append(mat[i][j])
                i += 1
                j -= 1
            if k%2:
                ans += t
            else:
                t.reverse()
                ans += t

        return ans
            
